### ISSUE TITLE
Optional member access for lists

### ISSUE THREAD

**Issue by Krzmbrzl** [microsoft/pyright#3300]
**Describe the bug**
Accessing an optional member stored in a list unconditionally errors, even if one has checked for ` None` explicitly.

**To Reproduce**
Use the following snippet and let pyright evaluate it:
```python
from typing import List
from typing import Optional

test: List[Optional[str]] = []

for i in range(len(test)):
    if not test[i] is None:
        print(test[i].find("bla"))
```
See that pyright complains about the access of `find` as it claims that `test[i]` could be `None`.

**Expected behavior**
Explicitly checking the element for `None` should get rid of this error.

**VS Code extension or command-line**
Neither - I'm using it as a vim COC-plugin: https://github.com/fannheyward/coc-pyright

**Comment by erictraut** [comment]
This is intended behavior. Type narrowing is not performed on subscript expressions with a dynamic subscript. It is performed when the subscript is a literal integer value, such as `test[1]`. The problem with dynamic subscripts is that a type checker cannot guarantee safety because lists (and other classes that accept subscripts) can be mutated in ways that are not visible to the type checker. Other Python type checkers (including mypy) and type checkers in other languages make the same assumption.

The correct way to handle this in a type-safe manner is to assign the value to a local variable, which is guaranteed to to change.

```python
for i in range(len(test)):
    if not (x := test[i]) is None:
        print(x.find("bla"))
```
